src/System.CommandLine.Hosting/HostingExtensions.cs

```diff
@@ -0,0 +1,48 @@
+﻿using System.CommandLine.Builder;
+using System.CommandLine.Invocation;
+using System.Linq;
+
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace System.CommandLine.Hosting
+{
+    public static class HostingExtensions
+    {
+        public static CommandLineBuilder UseHost(this CommandLineBuilder builder,
+            Func<string[], IHostBuilder> hostBuilderFactory,
+            Action<IHostBuilder> configureHost = null) =>
+            builder.UseMiddleware(async (invocation, next) =>
+            {
+                var argsRemaining = invocation.ParseResult.UnmatchedTokens.ToArray();
```

Comment:
>argsRemaining [](start = 20, length = 13)

What's an example of what these args would contain?

If these are coming from the command line input and filtered through the parser configuration, I wonder if this might be unpredictable.

Reply:
As shown in the added test case shown below unmatched tokens can be passed on to the HostConfiguration. If you include the `Microsoft.Extensions.Configuration.CommandLine` package.

https://github.com/dotnet/command-line-api/blob/9f1fa174542e5c6296bb7570f42cd1fb12df4a98/src/System.CommandLine.Hosting.Tests/HostingTests.cs#L91-L127

For example, you can thus pass `--environment Development` to set the Hosting Environment to `Development`. ASP.NET uses this to configure Kestrel from command-line arguments. However, the `AddCommandLine()` extension is really simple and only supports Key-Value-Pairs, so `System.CommandLine` is MUCH nicer to actual command-line argument parsing :)

The default on the `Command` constructor is `treatUnmatchedTokensAsError: true`, so normally the `UnmatchedTokens` property in `ParseResult` will be empty causing the host configuration to remain unaffected.

It is also worth pointing out that the normal `HostBuilder` constructor does not accept command-line arguments. That was only added in the 3.0 preview with the `Host.CreateDefaultBuilder` method. If have added test-cases showing how the unmatched command-line arguments can be passed into the Host configuration.

Reply:
There are a couple of other approaches we might want to explore here. Requiring someone to relax what is usually an error check seems restrictive.

One option is to use unparsed tokens: https://github.com/dotnet/command-line-api/blob/4c7695193bb79d649cd3711bb00639a0c2c5b458/src/System.CommandLine.Tests/ParserTests.cs#L160 Similar to the `UnmatchedTokens` though this does potentially conflict with someone's preferences for their app's command line interface.

An approach that wouldn't constrain the other parts of someone's interface would be to use directives: https://github.com/dotnet/command-line-api/wiki/Syntax-Concepts-and-Parser#directives. The wiki doesn't cover the fact that directives are parameterizable. In the following example, "config" is the directive name. Everything after the colon is the directive value. Multiple directives having the same name will aggregate the values into a collection.

```console
> myapp [config:key=value]           --int-option 123
         ^----^ ^-------^
```

Reply:
@jonsequitur Oh! I totally agree! I will change it it to use `UnparsedTokens` instead.

Using the a `config` directive sounds like a great idea. You could already do it yourself right now, by getting the `InvocationContext` from the `HostBuilderContext`. But making that first-class would be so much nicer and not break anyone.

I'll add that in ASAP :)